e2e/e2e_test.go

```diff
@@ -0,0 +1,301 @@
+				t.Fatalf("failed to set up cluster: %s", err)
+			}
+			defer func() {
+				if err := runScript(extraEnvs, "destroy_cluster.sh"); err != nil {
+					t.Errorf("failed to destroy cluster (post-test): %s", err)
+				}
+			}()
+
+			cs, err := kubeClient(kubeConfFile)
+			if err != nil {
+				t.Fatalf("failed to create Kubernetes client: %s", err)
+			}
+
+			// Check that nodes become ready.
+			l.Println("Polling for node readiness")
+			var (
+				gotNodes      []corev1.Node
+				numReadyNodes int
+			)
+			start := time.Now()
+			if err := wait.Poll(5*time.Second, 6*time.Minute, func() (bool, error) {
+				nl, err := cs.Core().Nodes().List(metav1.ListOptions{LabelSelector: "kubernetes.io/role=node"})
+				if err != nil {
+					return false, err
+				}
+
+				gotNodes = nl.Items
+				numReadyNodes = 0
+				for _, node := range gotNodes {
+					for _, cond := range node.Status.Conditions {
+						if cond.Type == corev1.NodeReady && cond.Status == corev1.ConditionTrue {
+							if _, ok := node.Labels[doLabel]; ok {
+								numReadyNodes++
+							}
+						}
+					}
+				}
+
+				l.Printf("Found %d/%d ready node(s)", numReadyNodes, numWantNodes)
+				return numReadyNodes == numWantNodes, nil
+			}); err != nil {
+				t.Fatalf("got %d ready node(s), want %d: %s\nnnodes: %v", numReadyNodes, numWantNodes, err, spew.Sdump(gotNodes))
+			}
+			l.Printf("Took %v for nodes to become ready\n", time.Since(start))
+
+			// Check that load balancer is working.
+
+			// Install example pod to load-balance to.
+			appName := "app"
+			pod := corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: appName,
+					Labels: map[string]string{
+						"app": appName,
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "nginx",
+							Image: "nginx",
+							Ports: []corev1.ContainerPort{
+								{
+									ContainerPort: 80,
+								},
+							},
+						},
+					},
+				},
+			}
+
+			if _, err := cs.CoreV1().Pods(corev1.NamespaceDefault).Create(&pod); err != nil {
+				t.Fatalf("failed to create example pod: %s", err)
+			}
+
+			// Wait for example pod to become ready.
+			l.Println("Polling for pod readiness")
+			start = time.Now()
+			var appPod *corev1.Pod
+			if err := wait.Poll(1*time.Second, 1*time.Minute, func() (bool, error) {
+				pod, err := cs.CoreV1().Pods(corev1.NamespaceDefault).Get(appName, metav1.GetOptions{})
+				if err != nil {
+					if kerrors.IsNotFound(err) {
+						return false, nil
+					}
+					return false, err
+				}
+				appPod = pod
+				for _, cond := range appPod.Status.Conditions {
+					if cond.Type == corev1.PodReady && cond.Status == corev1.ConditionTrue {
+						return true, nil
+					}
+				}
+				return false, nil
+			}); err != nil {
+				t.Fatalf("failed to observe ready example pod %q in time: %s\npod: %v", appName, err, appPod)
+			}
+			l.Printf("Took %v for pod to become ready\n", time.Since(start))
+
+			// Create service object.
+			svcName := "svc"
```

Comment:
Probably good to append these names with some random hash to avoid conflicts in future runs (in case they're not cleaned up in previous runs).

Reply:
Similar to pods, we should never collide as each test cluster is brought up for a single test case only.